libraries/standard/mqtt/src/mqtt_state.c

```diff
@@ -0,0 +1,636 @@
+            }
+            break;
+        default:
+            /* For a PUBLISH, we should not start from any other state. */
+            break;
+    }
+    return isValid;
+}
+
+static bool _validateTransitionAck( MQTTPublishState_t currentState,
+                                    MQTTPublishState_t newState )
+{
+    bool isValid = false;
+    switch( currentState )
+    {
+        case MQTTPubAckSend:
+            /* Incoming publish, QoS 1. */
+        case MQTTPubAckPending:
+            /* Outgoing publish, QoS 1. */
+            isValid = ( newState == MQTTPublishDone );
+            break;
+        case MQTTPubRecSend:
+            /* Incoming publish, QoS 2. */
+            isValid = ( newState == MQTTPubRelPending );
+            break;
+        case MQTTPubRelPending:
+            /* Incoming publish, QoS 2. */
+            isValid = ( newState == MQTTPubCompSend );
+            break;
+        case MQTTPubCompSend:
+            /* Incoming publish, QoS 2. */
+            isValid = ( newState == MQTTPublishDone );
+            break;
+        case MQTTPubRecPending:
+            /* Outgoing publish, Qos 2. */
+            isValid = ( newState == MQTTPubRelSend );
+            break;
+        case MQTTPubRelSend:
+            /* Outgoing publish, Qos 2. */
+            isValid = ( newState == MQTTPubCompPending );
+            break;
+        case MQTTPubCompPending:
+            /* Outgoing publish, Qos 2. */
+            isValid = ( newState == MQTTPublishDone );
+            break;
+        case MQTTPublishDone:
+            /* Done state should transition to invalid since it will be removed from the record. */
+        case MQTTPublishSend:
+            /* If an ack was sent/received we shouldn't have been in this state. */
+        case MQTTStateNull:
+            /* If an ack was sent/received the record should exist. */
+        default:
+            /* Invalid. */
+            break;
+    }
+
+    return isValid;
+}
+
+static bool _isPublishOutgoing( MQTTPubAckType_t packetType,
+                                MQTTStateOperation_t opType )
+{
+    bool isOutgoing = false;
+    switch( packetType )
+    {
+        case MQTTPuback:
+            isOutgoing = ( opType == MQTT_RECEIVE );
+            break;
+        case MQTTPubrec:
+            isOutgoing = ( opType == MQTT_RECEIVE );
+            break;
+        case MQTTPubrel:
+            isOutgoing = ( opType == MQTT_SEND );
+            break;
+        case MQTTPubcomp:
+            isOutgoing = ( opType == MQTT_RECEIVE );
+            break;
+        default:
+            /* No other ack type. */
+            break;
+    }
+    return isOutgoing;
+}
+
+static size_t _findInRecord( const MQTTPubAckInfo_t * records,
+                             size_t recordCount,
+                             uint16_t packetId,
+                             MQTTQoS_t * pQos,
+                             MQTTPublishState_t * pCurrentState )
+{
+    size_t index = 0;
```

Comment:
Can you add assert checks for input pointer parameters? 
Asserts are a expressing function's assumptions about the input parameters. As these functions are internal to the library, we have control over their calls, thus, the function can carry the assumption that the input parameters are valid (instead of adding error handling logic for invalid parameters )

Reply:
I was unsure if we wanted to include assert.h since we're not using it anywhere, so I had left it alone. I'll bring it up and will include it if desired.